src/lib/OS/MemRegion.cc

```diff
@@ -83,39 +76,47 @@ uint64_t MemRegion::updateBrkRegion(uint64_t newBrk) {
   return brk_;
 }
 
-uint64_t MemRegion::addVma(VMA* vma, uint64_t startAddr) {
-  VMA* curr = vm_head_;
-  size_t size = vma->vmSize_;
-  // When starAddr is not 0, search for an available address range
-  // that can hold the new VMA with a starting address that is greater than or
-  // equal to the specified startAddr. The following algorithm retrieves the
-  // last existing VMA object before the address range so that the new VMA can
-  // be linked between two existing VMAs. If no available address range is
-  // found, then the new VMA is allocated at the end of the VMA list.
-  if (startAddr && vm_size_ > 0) {
-    while (curr->vmNext_ != nullptr) {
-      if (curr->vmEnd_ <= startAddr &&
-          curr->vmNext_->vmStart_ >= (startAddr + size)) {
+uint64_t MemRegion::addVma(VMA vma, uint64_t startAddr) {
+  size_t size = vma.vmSize_;
+  auto last = std::prev(VMAlist_->end(), 1);
+  bool allocated = false;
+
+  auto itr = VMAlist_->begin();
+  if (VMAlist_->size() > 0) {
+    // Check if the new VMA can be allocated between mmapStart and the first VMA
+    // in the VMA list.
+    uint64_t effectiveMmapStart = startAddr ? startAddr : mmapStart_;
+    uint64_t space = itr->vmStart_ - effectiveMmapStart;
+    if (effectiveMmapStart < itr->vmStart_ && space >= size) {
+      vma.vmStart_ = effectiveMmapStart;
+      vma.vmEnd_ = effectiveMmapStart + size;
+      VMAlist_->insert(itr, vma);
+      return vma.vmStart_;
+    }
+    // If the VMA list has multiple VMAs then starting from the beginning of the
+    // VMA list check if the new VMA can be allocated between two existing ones.
+    // If startAddr is 0 check all address ranges between existing VMAs.
+    // However, if startAddr is non-zero then only search for an available
+    // address range which either contains startAddr or has a starting address
+    // greater than startAddr. If no address range is found, then the new VMA is
```

Comment:
Is this ruling defined somewhere?

Reply:
Yes this is the mmap specification. Please [see](https://github.com/UoB-HPC/SimEng/pull/302#discussion_r1151658491)